Fix bug; Update test case; Update limiting scheme

hlokavarapu opened the pull request:
Found a bug in the recently merged PR #1554 . This patch resolves the bug and updates the test case accordingly. Also, the limiter is improved where we make certain that the interpolated particle property is in the range of the global maximum particle property and global minimum particle property values that is specified in the parameter file. Previously, we would use the local minimum and maximum particle property values in a given cell as the range.

tjhei commented on source/particle/interpolator/bilinear_least_squares.cc:
```diff
@@ -148,16 +154,89 @@ namespace aspect
+                                  Patterns::Bool (),
+                                  "Whether to apply a global particle property limiting scheme to the interpolated particle properties "
+                                  "using bilinear least squares schmeme. The 'Global particle property maximum' and 'Global particle "
+                                  "property minimum' must be defined in the parameter file if set to true. ");
```
remove trailing space

tjhei commented on source/particle/interpolator/bilinear_least_squares.cc:
```diff
@@ -148,16 +154,89 @@ namespace aspect
-              interpolated_value = min_value_for_particle_property;
+            if (use_global_valued_limter)
+              {
+                if (interpolated_value > global_maximum_particle_properties[property_index])
```
it would be easier to do ``value = std::min(value, maximum);`` etc.

gassmoeller commented:
/run-tests

gassmoeller commented on tests/particle_interpolator_bilinear_least_squares_solkz/screen-output:
```diff
@@ -1,19 +1,19 @@
-     Writing particle output:       output-particle_interpolator_bilinear_solkz/particles/particles-00000
-     Errors u_L1, p_L1, u_L2, p_L2: 2.396693e-07, 3.813761e-04, 3.798509e-07, 1.430329e-03
+     Writing particle output:       output-particle_interpolator_bilinear_least_squares_solkz/particles/particles-00000
+     Errors u_L1, p_L1, u_L2, p_L2: 1.768545e-08, 1.580101e-04, 3.863311e-08, 2.648933e-04
```
Is this the same test change that occured in #1700? Please check that your test actually uses the particle properties for the material model. If so, is this change of results by chance? Or is the difference between the particle and the compositional fields case exactly caused by the over-/undershoot in the interpolation scheme?

bangerth replied:
Yes, this looks like the same issue as before.

hlokavarapu replied:
After rebasing against master, I checked that the SolKz material model is reached at run time:
```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffde433700 (LWP 6411)]

Thread 1 "aspect" hit Breakpoint 2, aspect::InclusionBenchmark::SolKzCompositionalMaterial<2>::evaluate (this=<optimized out>, in=..., out=...) at /home/harsha_lv/ClionProjects/aspect/tests/particle_interpolator_bilinear_least_squares_solkz/../../benchmarks/solkz/compositional_fields/solkz_compositional_fields.h:24
24	          for (unsigned int i=0; i < in.position.size(); ++i)
(gdb) bt
#0  aspect::InclusionBenchmark::SolKzCompositionalMaterial<2>::evaluate (this=<optimized out>, in=..., out=...) at /home/harsha_lv/ClionProjects/aspect/tests/particle_interpolator_bilinear_least_squares_solkz/../../benchmarks/solkz/compositional_fields/solkz_compositional_fields.h:24
#1  0x000000000085c4e7 in aspect::AdiabaticConditions::InitialProfile<2>::initialize (this=<optimized out>) at /home/harsha_lv/ClionProjects/aspect/source/adiabatic_conditions/initial_profile.cc:137
#2  0x0000000000d9b0a5 in aspect::Simulator<2>::Simulator (this=0x7fffffffb5b0, mpi_communicator_=0x7fffed547680 <ompi_mpi_comm_world>, prm=...) at /home/harsha_lv/ClionProjects/aspect/source/simulator/core.cc:533
#3  0x0000000000990170 in main (argc=2, argv=0x7fffffffdc28) at /home/harsha_lv/ClionProjects/aspect/source/main.cc:603

```
The difference in the compositional values in the statistics file is mainly because I accidentally set the test case to utilize the cell average scheme. This patch corrects that mistake.

hlokavarapu commented:
I ran a convergence test of the benchmark and below are the results for a fixed number of particles per cell (16). Note that the difference in the compositional field values and that of the bilinear least squares interpolation scheme can be reduced by increasing the number of particles per cell.

![solkz](https://cloud.githubusercontent.com/assets/1869576/26320657/6ab1596c-3ed8-11e7-8ec4-64f98a975510.png)

gassmoeller commented on include/aspect/particle/interpolator/bilinear_least_squares.h:
```diff
@@ -54,6 +54,25 @@ namespace aspect
+        private:
+          bool use_global_valued_limter;
+          std::vector<double> global_maximum_particle_properties;
+          std::vector<double> global_minimum_particle_properties;
```
Please add a short comment to the three variables

gassmoeller commented on source/particle/interpolator/bilinear_least_squares.cc:
```diff
@@ -51,6 +54,18 @@ namespace aspect
+                        ExcMessage("Make sure that the size of 'Global maximum particle property' is equivalent to the number "
+                                   "of particle properties."));
+
+          }
```
Please move this whole block into the `parse_parameters` function. You will likely need to get the number of particle properties in a different way (they will not be initialized at that point), but something like the following should work:
```
const Postprocess::Particles<dim> *particle_postprocessor = this->template find_postprocessor<Postprocess::Particles<dim> >();

const unsigned int n_particle_properties = particle_postprocessor->get_particle_world()->get_property_manager->get_data_info()->n_components();
```

This will be both faster (this function is called a lot), and safer (usually we check input parameters right after reading them to not keep wrong values around).

hlokavarapu replied:
In calling the function
`this->template find_postprocessor<Postprocess::Particles<dim> >();` from `aspect::Particle::Interpolator::BilinearLeastSquares<dim>::properties_at_points()`, my compiler produces the error message

```
/home/harsha_lv/ClionProjects/aspect/include/aspect/simulator_access.h: In instantiation of ‘PostprocessorType* aspect::SimulatorAccess<dim>::find_postprocessor() const [with PostprocessorType = aspect::Postprocess::Particles<2>; int dim = 2]’:
/home/harsha_lv/ClionProjects/aspect/source/particle/interpolator/bilinear_least_squares.cc:44:135:   required from ‘std::vector<std::vector<double> > aspect::Particle::Interpolator::BilinearLeastSquares<dim>::properties_at_points(const std::multimap<std::pair<int, int>, aspect::Particle::Particle<dim> >&, const std::vector<dealii::Point<dim> >&, const dealii::ComponentMask&, const typename dealii::parallel::distributed::Triangulation<dim>::active_cell_iterator&) const [with int dim = 2; typename dealii::parallel::distributed::Triangulation<dim>::active_cell_iterator = dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> >]’
/home/harsha_lv/ClionProjects/aspect/source/particle/interpolator/bilinear_least_squares.cc:255:7:   required from here
/home/harsha_lv/ClionProjects/aspect/include/aspect/simulator_access.h:684:90: error: invalid use of incomplete type ‘const class aspect::Simulator<2>’
     return simulator->postprocess_manager.template find_postprocessor<PostprocessorType>();

```
The BilinearLeastSquares class inherits from the Particle::Interpolator::Interface class and SimulatorAccess class. Does one need to treat the registration of the BilinearLeastSquares class in a special way to use the function `find_postprocessor`?? If you have any quick leads, that would be helpful. Thanks.

gassmoeller replied:
Yes, you will need to add the following includes (if they are not already included in the .cc or .h file) to make that work:
```
#include <aspect/particle/world.h>
#include <aspect/simulator.h>
```

gassmoeller commented on source/particle/interpolator/bilinear_least_squares.cc:
```diff
@@ -148,16 +154,88 @@ namespace aspect
-              interpolated_value = max_value_for_particle_property;
-            else if (interpolated_value < min_value_for_particle_property)
-              interpolated_value = min_value_for_particle_property;
+            if (use_global_valued_limter)
```
limiter?

gassmoeller commented on include/aspect/particle/interpolator/bilinear_least_squares.h:
```diff
@@ -54,6 +54,25 @@ namespace aspect
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          bool use_global_valued_limter;
```
limiter?

hlokavarapu replied:
I am calling this a limiter, although a simplified version, because this is in the same spirit of the parameters

```
Use limiter for discontinuous composition solution
Global temperature maximum
Global temperature minimum
```
which are part of the Discretization subsection. In the near future, we plan on reusing the already implemented limiting scheme for temperature and composition.

gassmoeller replied:
Sorry, I should have been more explicit: There is a typo in the variable name `limter` -> `limiter` ... I did not mean to question the name itself :smile:

gassmoeller commented on source/particle/interpolator/bilinear_least_squares.cc:
```diff
@@ -148,16 +154,88 @@ namespace aspect
+                                   "The maximum global particle property values that will be used as a "
+                                   "limiter for the bilinear least squares interpolation. The number of the input "
+                                   "'Global particle property maximum' values separated by ',' has to be "
+                                   "the same as the number of the interpolated particle properties.");
```
please remove the `interpolated` from the last line. We check whether the number of values is equal to the total number of properties, because we do not know how many will be interpolated.

gassmoeller commented on source/particle/interpolator/bilinear_least_squares.cc:
```diff
@@ -148,16 +154,88 @@ namespace aspect
+                                   "The minimum global particle property that will be used as a "
+                                   "limiter for the bilinear least squares interpolation. The number of the input "
+                                   "'Global particle property minimum' values separated by ',' has to be "
+                                   "the same as the number of the interpolated particle properties.");
```
same here

gassmoeller commented on source/particle/interpolator/bilinear_least_squares.cc:
```diff
@@ -148,16 +154,88 @@ namespace aspect
+                prm.declare_entry("Use limiter", "false",
+                                  Patterns::Bool (),
+                                  "Whether to apply a global particle property limiting scheme to the interpolated "
+                                  "particle properties using bilinear least squares schmeme. The 'Global particle "
```
I would shorten the first sentence as follows: `Whether to apply a global limiting scheme to the interpolated particle properties.` All other information is already available to the user.

gassmoeller commented:
Looks good, thanks! Please rebase one more time to resolve the conflict, then this is ready to go.

hlokavarapu commented:
I have rebased this PR and if the tester is happy, the PR should be ready to be reviewed for another iteration. Note that I have also squashed my commits where the only additional change I have committed is removing `#include <aspect/utlilites.h>" which was unnecessary in the `bilinear_least_squares.cc`.
